src/Compilers/CSharp/Portable/Symbols/Source/SourceNamedTypeSymbol_Extension.cs

```diff
@@ -63,6 +64,321 @@ internal override string ExtensionName
+        /// It is meant be to hashed to produce the content-based name for the extension grouping type.
+        /// </summary>
+        internal string ComputeExtensionGroupingRawName()
+        {
+            var pooledBuilder = PooledStringBuilder.GetInstance();
+            var builder = pooledBuilder.Builder;
+            builder.Append("extension");
+
+            if (this.IsGenericType)
+            {
+                builder.Append("<");
+
+                foreach (var typeParameter in this.TypeParameters)
+                {
+                    if (typeParameter.Ordinal > 0)
+                    {
+                        builder.Append(", ");
+                    }
+
+                    appendTypeParameter(typeParameter, builder, isDefinition: true);
+                }
+
+                builder.Append(">");
+            }
+
+            builder.Append("(");
+            if (this.ExtensionParameter is { } extensionParameter)
+            {
+                appendType(extensionParameter.Type, builder);
+            }
+
+            builder.Append(")");
+
+            return pooledBuilder.ToStringAndFree();
+
+            static void appendType(TypeSymbol type, StringBuilder builder)
+            {
+                if (type is NamedTypeSymbol namedType)
+                {
+                    appendNamedType(type, builder, namedType);
+                }
+                else if (type is TypeParameterSymbol typeParameter)
+                {
+                    appendTypeParameter(typeParameter, builder, isDefinition: false);
+                }
+                else if (type is ArrayTypeSymbol array)
+                {
+                    appendArrayType(array, builder);
+                }
+                else if (type is PointerTypeSymbol pointer)
+                {
+                    appendType(pointer.PointedAtType, builder);
+                    builder.Append('*');
+                }
+                else if (type is FunctionPointerTypeSymbol functionPointer)
+                {
+                    appendFunctionPointerType(functionPointer, builder);
+                }
+                else
+                {
+                    throw ExceptionUtilities.UnexpectedValue(type);
+                }
+            }
+
+            static void appendNamedType(TypeSymbol type, StringBuilder builder, NamedTypeSymbol namedType)
+            {
+                Debug.Assert(type.CustomModifierCount() == 0);
+
+                if (namedType.SpecialType == SpecialType.System_Void)
+                {
+                    builder.Append("void");
+                    return;
+                }
+
+                if (namedType.IsReferenceType)
+                {
+                    builder.Append("class ");
+                }
+                else if (namedType.IsValueType)
+                {
+                    builder.Append("valuetype ");
+                }
```

Comment:
I think we can drop these prefixes otherwise we are going to allow class and struct to conflict by full name.

Reply:
> The reason I included those keywords is to conform to valid IL.

I do not think it is our goal to produce a proper IL. We certainly use IL grammar as an inspiration, but I think we can drop some aspects of it that are not part of our specification of what should be included into the content-based name of the extension grouping type.